src/compiler/utils/string_with_sourcemap.ts

```diff
@@ -0,0 +1,273 @@
+import { DecodedSourceMap, RawSourceMap, SourceMapLoader } from '@ampproject/remapping/dist/types/types';
+import remapping from '@ampproject/remapping';
+import { SourceMap } from 'magic-string';
+
+type SourceLocation = {
+	line: number;
+	column: number;
+};
+
+function last_line_length(s: string) {
+	return s.length - s.lastIndexOf('\n') - 1;
+}
+
+// mutate map in-place
+export function sourcemap_add_offset(
+	map: DecodedSourceMap, offset: SourceLocation
+) {
+	if (map.mappings.length == 0) return map;
+	// shift columns in first line
+	const segment_list = map.mappings[0];
+	for (let segment = 0; segment < segment_list.length; segment++) {
+		const seg = segment_list[segment];
+		if (seg[3]) seg[3] += offset.column;
+	}
+	// shift lines
+	for (let line = 0; line < map.mappings.length; line++) {
+		const segment_list = map.mappings[line];
+		for (let segment = 0; segment < segment_list.length; segment++) {
+			const seg = segment_list[segment];
+			if (seg[2]) seg[2] += offset.line;
+		}
+	}
+}
+
+function merge_tables<T>(this_table: T[], other_table: T[]): [T[], number[], boolean, boolean] {
+	const new_table = this_table.slice();
+	const idx_map = [];
+	other_table = other_table || [];
+	let val_changed = false;
+	for (const [other_idx, other_val] of other_table.entries()) {
+		const this_idx = this_table.indexOf(other_val);
+		if (this_idx >= 0) {
+			idx_map[other_idx] = this_idx;
+		} else {
+			const new_idx = new_table.length;
+			new_table[new_idx] = other_val;
+			idx_map[other_idx] = new_idx;
+			val_changed = true;
+		}
+	}
+	let idx_changed = val_changed;
+	if (val_changed) {
+		if (idx_map.find((val, idx) => val != idx) === undefined) {
+			// idx_map is identity map [0, 1, 2, 3, 4, ....]
+			idx_changed = false;
+		}
+	}
+	return [new_table, idx_map, val_changed, idx_changed];
+}
+
+function pushArray<T>(_this: T[], other: T[]) {
```

Comment:
can you replace this with array's `concat` method?

Reply:
I think this was done for perf. concat is much slower. But maybe I can ditch this method and replace with array1.push(...array2)

Reply:
I can't replace with array1.push(...array2) since in some cases we might hit the javascript function argument limit
> The JavaScriptCore engine has hard-coded argument limit of 65536.

Reply:
ok, maybe we just leave it as is, but how about adding a comment explaining it's done this way for performance

Reply:
my bad (missing comment) .. this is the fastest solution to concat large arrays

`a1.push(...a2)` is limited by "call stack size" 
which on node/chrome is around 100K. on firefox around 500K

`a1 = a1.concat(a2)` is slower cos "immutable data"

these just look pretty ..

related: https://github.com/sveltejs/svelte/issues/4694